### ISSUE TITLE
GC experimentation

### ISSUE THREAD

**Issue by kripken** [AssemblyScript/assemblyscript#89]
The [wasm GC](https://github.com/WebAssembly/gc/blob/master/proposals/gc/Overview.md) spec is making progress, initial experimental implementations are starting to be worked on in VMs, and there is experimentation [in AssemblyScript too](https://github.com/AssemblyScript/assemblyscript/tree/gc-integration). It would be great to be able to connect all these things! That is, to know what the earliest testable GC things in VMs are going to be, so that we can emit them from AssemblyScript and then see that code running in a VM. Then we'd be testing the whole `toolchain->VM` pipeline.

A possible vague plan:
 * Figure out the right GC things to target, that will be testable soon.
 * Implement those things in Binaryen, defining new APIs as needed.
 * Use those new APIs in AssemblyScript.

(Personally I'd love to help out with this, on the Binaryen parts.)

**Comment by kripken** [comment]
One question I have about this, for @dcodeIO: perhaps you can give an overview of what the GC plans are here? (Sorry if this is on the wiki, I couldn't see it.) Specifically, how do you intend to use GC - is it for external JS things, or also for AssemblyScript objects themselves, or something else?

And will it be an optional flag that any AssemblyScript code can use, or will source code be written somewhat differently for it?

**Comment by dcodeIO** [comment]
> perhaps you can give an overview of what the GC plans are here?

Plan so far was to implement a makeshift GC as a placeholder while the GC spec is still in the works, but given that there is something we can test with on the horizon, I am considering skipping a makeshift GC and instead focusing on integrating with WASM GC right away, eventually making it the primary implementation in AssemblyScript (for internal and external objects). This is going to change the way how class instances and arrays are allocated entirely, of course, but that's expected. If we have to break / throw away the existing allocation stuff, that's fine for me.

> And will it be an optional flag that any AssemblyScript code can use, or will source code be written somewhat differently for it?

Hence: Probably not behind a flag, but the new default right away, simply living in its own branch until support lands in VMs.

**On the current state of things:**

Currently, whenever a class is instantiated, a suitably-sized slab of memory is allocated and its fields layed out within. The compiler does the layout and injects allocations into constructors. With WASM GC, there'd be no such allocation anymore (most likely not even a userspace allocator by default), and the class will become a WASM GC tuple/struct using its fields as elements.

Similarly, whenever an array is instantiated, a suitably-sized slab of memory is allocated that becomes the backing ArrayBuffer of the Array. The array itself is a class with a field pointing to the backing buffer. With WASM GC, the ArrayBuffer will become a WASM GC array (with possibly packed elements to account for i8, i16), and the Array tuple/struct will reference it - unless there is a way to mix GC tuples and arrays (doesn't seem so at this point), which would safe us one level of indirection.

The current design moves a lot of the implementation details to the standard library, implemented in TypeScript, that explicitly includes code to do the necessary loads/stores to/from linear memory (using methods annotated as operator overloads), for example when an indexed access on an array is performed. For fields, the compiler currently generates loads with a constant offsets using the `this` pointer according to the previously computed layout. This would change with WASM GC, in that the compiler would now generate the respective tuple/struct/array types and the standard library would just sit on top of it, effectively moving this logic from the standard library into the compiler.

Maybe a little shopping list for what would be necessary to make the transition:

**Classes**

* Tuple types with typed elements (`(type $XY (struct ... ))`), represented by a typed runtime reference (`(ref $XY)`)
* Instructions to load (`(load_field ...)`) and store fields (`(store_field ...)`) using the runtime reference
* Instruction to instantiate a new instance (`(new_default_struct ...)`)

**Arrays**

* Array types with (packed) elements (`(type $XY (array ...))`) with flexible size, represented by a typed runtime reference (`(ref $XY)`)
* Instructions to load (`(load_elem ...)`) and store elements (`(store_elem ...)`), by index, using the runtime reference
* Instruction to instantiate a new instance (`(new_default_array ...)`)
* Here it's important that arrays have flexible sizes (no way to annotate fixed sizes in TypeScript), but this is mentioned as the default in the spec as far as I can see

With the above in place it should be possible to start with an initial implementation (throw away unnecessary parts of the standard library, re-implement in the compiler, i.e. indexed array access, property accesses on class instances and changing linear allocation to `new_default_struct` etc.). A good follow-up might be to look at extending classes with a way to downcast a reference (here e.g. the `super` keyword in TS; interfaces aren't yet implemented in AssemblyScript yet, so we can start with just the basics).

I hope I didn't miss something obvious here, but if I did or failed to clarify some parts, feel free to mention it :)

**Comment by kripken** [comment]
Thanks @dcodeIO !

From talking with @lukewagner (Luke, correct me if I got something wrong), the first step here for VMs is to implement the "reference types proposal", which define some [new types](https://webassembly.github.io/reference-types/core/exec/instructions.html#reference-instructions) and [new instructions](https://webassembly.github.io/reference-types/core/exec/instructions.html#table-instructions). This may already be ready for experimentation? However, it is **not** enough for the AssemblyScript stuff described above like creating a new object and getting and setting fields, instead, if I understand the spec it just adds

 * A few new fixed types like "anyfunction" (which is any function instead of "anyref" which is anything).
 * Reference instructions which can compare refs.
 * Table operations like read and write refs to/from tables.

In other words this is a small step towards full GC support, but only provides table and related operations.

Would reference types be interesting to experiment with in AssemblyScript? Should be quick to add to Binaryen if so. Or would we need to wait for full GC support?

**Comment by kripken** [comment]
Yeah, I don't really see the table instruction use case either. Overall, it doesn't seem like that much is possible with just the reference-types proposal, and we need to wait for full GC. @lukewagner, am I missing something?

**Comment by lukewagner** [comment]
The use case for tables-of-refs is when you want to logically store a reference in linear memory.  Then you stick it into a table and store its `i32` index in linear memory.  This requires having a lifetime for the table element so that you can free up the element and avoid leaking the object forever.  E.g., see how [Rust's wasm-bindgen](https://github.com/rustwasm/wasm-bindgen/blob/master/DESIGN.md#polyfill-for-js-objects-in-wasm) uses it.

For AssemblyScript, since the goal (I assume?) is for all objects to be allocated in the GC heap (not linear memory), there wouldn't be any need for this rooting so I agree there wouldn't be any need.

So yeah, I'd start with an `anyref` type that could point to any typed *or untyped* JS object.  One interesting extension which I think would be useful in the long-term in any case would be to have a syntax for doing dynamic dictionary object access on `anyref`s.  Initially this would be implemented by:
* creating an `anyref` holding the property name string (either once up-front for `obj.prop` or dynamically for `obj[prop]`)
* importing and calling `Reflect.get` passing `obj` and `prop`

This impl strategy will actually be somewhat slow (compared to native JS property access), but we can optimize it in various ways over time (both in engines, and possibly in the standard via a special get-prop Host Binding).

**Comment by dcodeIO** [comment]
@MaxGraey I think that spec'ed-features should become the default (best interoperability between different languages, smallest binaries), but we can of course just rename the current native implementation of arrays to `NativeArray` for example and make it configurable on compilation similar to how `--use Math=NativeMath` works (in case of array, there'd be `JSArray` and `NativeArray` then). The challenge there will be to make both kinds of arrays largely interoperable (see paragraph below), of course. Wdyt?

@lukewagner Thanks for the clarification, makes sense. So tables will actually be important to make structures that live in linear memory interoperable with references that don't. Here, for example, a `NativeArray<anyref>` (see paragraph above, or later also `NativeArray<JSArray>`) then stores the table indexes in its backing ArrayBuffer as `i32`s, and indexed access on such on array would have to wrap the `load` in a `get_table`. That'll be interesting to implement :)

**Comment by dcodeIO** [comment]
FYI: There is now support for [sign-extension-ops](https://github.com/WebAssembly/sign-extension-ops) and [mutable-global](https://github.com/WebAssembly/mutable-global) behind feature flags. With this mechanism present, the same can be done for reference-types without the need to keep it in its own branch :)

**Comment by vgrichina** [comment]
What is current status of this? Are there any VMs already implementing GC spec? Is there any roadmap on when spec will be finalized?

**Comment by dcodeIO** [comment]
Maybe a quick update on this for everyone stumbling upon this issue: From what I can tell, there is no sign of WASM GC on the horizon yet. However, recent commits to master implement a custom MM/GC combination based on reference counting for the time being, as [described here](https://docs.assemblyscript.org/details/runtime).

**Comment by MaxGraey** [comment]
Just leave this paper which describe how we could further improve Reference Counting
[Lazy_Cyclic_Reference_Counting.pdf](https://github.com/AssemblyScript/assemblyscript/files/3822659/Lazy_Cyclic_Reference_Counting.pdf)

**Comment by MaxGraey** [comment]
Paper investigated Immix approach for RC:
https://www.microsoft.com/en-us/research/wp-content/uploads/2016/02/conservative-gc-oopsla-2014.pdf

**Comment by MaxGraey** [comment]
Another paper about RC Immix
[ThesisShahriyar2015.pdf](https://github.com/AssemblyScript/assemblyscript/files/3984269/ThesisShahriyar2015.pdf)

**Comment by dcodeIO** [comment]
With the GC subgroup having regular meetings meanwhile and initial prototype implementations being pursued in WABT and V8, I've created a follow-up for the necessary Binaryen support here: https://github.com/WebAssembly/binaryen/issues/2935

It's likely that it's still a bit too early, so don't hold your breath.

**Comment by tlively** [comment]
If anyone from the AssemblyScript community would be interested in working to implement experimental GC support in Binaryen, I would be happy to help out with guidance and code reviews.

**Comment by MaxGraey** [comment]
I guess we could close this. Or keep open until wasm GC?

**Comment by tlively** [comment]
Yes, this work is well under way in Binaryen and supported in V8. I think this general issue can probably be closed.
